### PR TITLE
Implement EXLA.Backend.to_batched_list/3

### LINKED ISSUE #774: Implement Nx.to_batched_list for EXLA

### PR DESCRIPTION
**Pull request by jonatanklosko**
Closes #774.

### CODE REVIEW

File: exla/lib/exla/backend.ex [hunk 1]
```diff
@@ -67,9 +67,35 @@ defmodule EXLA.Backend do
...
+    remainder = rem(axis_size, batch_size)
+    num_full_batches = div(axis_size, batch_size)
+
+    to_add = if remainder != 0, do: batch_size - remainder, else: 0
+
+    {num_batches, expr_fun} =
+      if to_add != 0 and leftover == :repeat do
+        {num_full_batches + 1,
+         fn tensor, start_idx ->
+           wrapped_tensor = Nx.concatenate([tensor, Nx.slice_along_axis(tensor, 0, to_add)])
```
Comment (josevalim) [inline exla/lib/exla/backend.ex:85]: We only need to do the concatenation in the last operation. So I recommend jitting the same for all slices, except the last! Or writing conditional code inside defn.

Comment (jonatanklosko) [reply exla/lib/exla/backend.ex:85]: Oh, we would do this on every call, not just once, good point!

**Comment by jonatanklosko** [comment]
Updated! :cat:

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by noozo** [comment]
Thought it was possibly related to https://github.com/elixir-nx/nx/issues/776, but not really. Just trying re-running the credit_card_fraud axon example with the updated code:

```
defp deps do
    [
      {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
       {:exla, github: "elixir-nx/nx", sparse: "exla", override: true},
      # {:torchx, "~> 0.2"},
      {:nx, "~> 0.2.0", override: true},
      {:explorer, "~> 0.1.1"}
    ]
end
```

Indeed it no longer requires backend_transfer (or backend_copy), but still blows up with:

```
Epoch: 29, Batch: 100, loss: 0.0000009
** (Axon.CompilerError) error while building prediction for sigmoid:
** (Axon.CompilerError) error while building prediction for dense:
** (Axon.CompilerError) error while building prediction for dropout:
** (Axon.CompilerError) error while building prediction for relu:
** (Axon.CompilerError) error while building prediction for dense:
** (Axon.CompilerError) error while building prediction for relu:
** (Axon.CompilerError) error while building prediction for dense:

** (RuntimeError) cannot invoke Nx function because it relies on two incompatible tensor implementations: Nx.Defn.Expr and EXLA.Backend. You may need to call Nx.backend_transfer/1 (or Nx.backend_copy/1) on one or both of them to transfer them to a common implementation

    (nx 0.2.0) lib/nx/shared.ex:421: Nx.Shared.pick_struct/2
    (nx 0.2.0) lib/nx.ex:8030: Nx.dot/6
    (axon 0.1.0-dev) lib/axon/layers.ex:109: Axon.Layers."__defn:dense__"/3
    (axon 0.1.0-dev) lib/axon/compiler.ex:374: Axon.Compiler.recur_predict_fun/5
    (axon 0.1.0-dev) lib/axon/compiler.ex:182: Axon.Compiler.to_predict_fun/5
    (axon 0.1.0-dev) lib/axon/compiler.ex:288: Axon.Compiler.recur_predict_fun/5
    (axon 0.1.0-dev) lib/axon/compiler.ex:182: Axon.Compiler.to_predict_fun/5
    (elixir 1.13.2) lib/enum.ex:1715: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
```

Probably two separate issues though, but thought i mentioned it here just in case :)

**Comment by jonatanklosko** [comment]
@noozo yeah #776 is something else, but hopefully we can fix it soon!

**Pull request merged**